### PR TITLE
Custom deploy procedure

### PR DESCRIPTION
**Pull request by dtantsur**
See https://github.com/metal3-io/metal3-docs/blob/master/design/deploy-steps.md

### CODE REVIEW

File: apis/metal3.io/v1alpha1/baremetalhost_types.go
Comment (redixin) [inline apis/metal3.io/v1alpha1/baremetalhost_types.go:425]: What about args?

Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:425]: I was requested to omit args on the review of the design document. They can be added later if we have a good case (and an idea how to wire them in the API).

Comment (redixin) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:425]: What do you mean `a good case`? Passing args to the method isn't a good case?

Also `Args` is required field in DeployStep https://github.com/gophercloud/gophercloud/blob/v0.17.0/openstack/baremetal/v1/nodes/requests.go#L427

Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:425]: This implementation passes empty args (btw I'm not sure why we require them in ironic).

Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:425]: By a good case I mean a use case for passing args. What I'm doing does not require non-empty args.

Comment (redixin) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:425]: E.g when someone want to use ansible to deploy.

Your comment:

> #180 has been made much more generic, and you may be able to reuse its interface for your goals. E.g. you allow `CustomDeploy.Method = "ansible"` to designate the ansible deploy. Then you can use metaData for passing additional information.

https://github.com/metal3-io/metal3-docs/pull/181#discussion_r629368593

But regardless this comment, passing args to methods is pretty standard approach and methods without args are pretty much useless.

I understand that doc is merged and only `Method` is mentioned there, but am I only one who confused here?

"Ironic api allows to pass args but let's not use it."

I also can't see any discussion about removing args here: https://github.com/metal3-io/metal3-docs/pull/180

What is wrong with args?

Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:425]: The previous discussion is https://github.com/metal3-io/metal3-docs/pull/180#discussion_r626924686. The problem was building API that accepts arbitrary JSON data. We already have `MetaData` and `UserData` for that, although it may not work for the ansible case. In any case, I suggest you rebase your older design on top of https://github.com/metal3-io/metal3-docs/pull/180 and discuss it there. This patch implements the approved design.

**Review by elfosardo — APPROVED** [review]

File: apis/metal3.io/v1alpha1/baremetalhost_types.go
Comment (hardys) [inline apis/metal3.io/v1alpha1/baremetalhost_types.go:888]: I'm not sure we really want the new spec field to trigger reprovisioning?

Currently the only field which does that is the image, so as an existing user I might be surprised if I changed some value in the custom deploy field and all my hosts got rebuilt?

I'm thinking about consistency e.g with userData/networkData etc where those fields (or rather the secrets they reference) could change post-deploy, but we don't rebuild even though they influence the deployed OS - this is similar even though it won't reference a potentially-shared secret?

File: controllers/metal3.io/baremetalhost_controller.go
Comment (hardys) [inline controllers/metal3.io/baremetalhost_controller.go:855]: How can this happen - shouldn't the Status.Provisioning.Image always be removed by deprovisioning?

If this is a corner case found writing the test, I'd break it out into a separate commit as I'm currently not seeing how it's related to this change?

Comment (dtantsur) [reply controllers/metal3.io/baremetalhost_controller.go:855]: > shouldn't the Status.Provisioning.Image always be removed by deprovisioning?

This is provisioning code. Previously we would never get here if Spec.Image is nil, now we can. The old code causes segfault if we use Image == nil and CustomDeploy != nil.

File: apis/metal3.io/v1alpha1/baremetalhost_types.go
Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:888]: If we don't do this, how will we unprovision? In case when Image == nil and CustomDeploy != nil, the only way to unprovision is to set CustomDeploy = nil.

Comment (hardys) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:888]: Perhaps I'm confused, can you outline the expected provision/deprovision flow here?

I expect a BMH to deprovision only when I remove (or replace) the image URL, so from a users perspective I'd expect the same behavior even if I have chosen to use a custom deploy ramdisk?

File: controllers/metal3.io/baremetalhost_controller.go
Comment (hardys) [reply controllers/metal3.io/baremetalhost_controller.go:855]: Again perhaps I'm missing something, can you explain in more detail why we can now hit the provisioning path with an image set in the status?  IIUC that should never happen, because we should have deprovisioned first.  /cc @zaneb

Comment (dtantsur) [reply controllers/metal3.io/baremetalhost_controller.go:855]: I'm not sure why you're asking me, that's the code that existed before my change. I just made sure it doesn't fail with segfault now that Spec.Image can be niI.

I assume it's this code that synchronizes Status with Spec after deployment, but I may be wrong.

File: apis/metal3.io/v1alpha1/baremetalhost_types.go
Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:888]: Yes, but you don't have an image now. What would you do to deprovision?

My assumption was that a user removes CustomDeploy from Spec, similarly to how Image removal causes deprovisioning now.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [reply controllers/metal3.io/baremetalhost_controller.go:855]: Looking more at the code. I can remove this check if you think it's safe to, but I'm not 100% sure myself. Previously this code path would never be hit with Spec.Image == nil.

Comment (hardys) [reply controllers/metal3.io/baremetalhost_controller.go:855]: I'm asking you because the `info.host.Status.Provisioning.Image = metal3v1alpha1.Image{}` is new - and AFAIK we should never clear the image in the status during provisioning, only during deprovisioning

So, I'd like to understand under what circumstances the state machine can allow us to get into this state, whether the issue is existing code or not :)

File: apis/metal3.io/v1alpha1/baremetalhost_types.go
Comment (hardys) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:888]: In the current flow, if the `spec.image` gets removed, that triggers deprovisioning - in the case where CustomDeploy is specified I'd expect the same behavior.

But I probably would *not* expect to trigger rebuild of the host, if I modify CustomDeploy on an already provisioned host - most other spec fields don't trigger rebuild, so I think I'd expect this to behave the same.

If someone removes CustomDeploy from the spec, I'd expect to revert to the existing behavior (use the default deploy steps and IPA kernel/ramdisk) - obviously there are likely cases where it doesn't make sense to remove CustomDeploy because that configuration is required for provisioning and deprovisioning, but that's up to the user?

File: controllers/metal3.io/baremetalhost_controller.go
Comment (hardys) [reply controllers/metal3.io/baremetalhost_controller.go:855]: To clarify, I'm saying we should only clear the image in [actionDeprovisioning](https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/baremetalhost_controller.go#L917) - this addition to `actionProvisioning` seems wrong, so I want to understand if/why it's needed.

File: apis/metal3.io/v1alpha1/baremetalhost_types.go
Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:888]: Sorry, I don't get it. How do you suggest we trigger reprovision if Image is nil already?

Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:888]: > use the default deploy steps and IPA kernel/ramdisk

We have untied this logic from custom image logic, so only the default steps. But then we end up with a provisioned node with Image==nil and CustomDeploy==nil?

Comment (hardys) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:888]: Ok we discussed this, and the reason for my confusion is that it's now possible to have `CustomDeploy!=nil`, but with `Image==nil` e.g the Image field is now optional (which is mentioned in the [design](https://github.com/metal3-io/metal3-docs/blob/master/design/deploy-steps.md))

That is unfortunate since it changes the existing API behavior, and potentially means more code corner cases since we have two APIs now that trigger provisioning/deprovisioning.

I know @zaneb had some suggestions on using the image URL instead, but since we accepted the design with this API I'm not blocking on it unless others think we need to revisit that discussion.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (andfasano) [inline controllers/metal3.io/baremetalhost_controller.go:601]: I guess a `.DeepCopy()` may be required here?

File: controllers/metal3.io/baremetalhost_controller_test.go
Comment (andfasano) [inline controllers/metal3.io/baremetalhost_controller_test.go:1040]: I'd find easier to understand this test if defined in the table driven mode, given the presence of multiple cases

File: controllers/metal3.io/baremetalhost_controller_test.go
Comment (andfasano) [inline controllers/metal3.io/baremetalhost_controller_test.go:1116]: Just to be sure: `host.Status.Provisioning.CustomDeploy.Method == "install_everything"`

Also, to be on the safe side, it could be worth adding a check on the state, ie `.. && host.Status.Provisioning.State == metal3v1alpha1.StateProvisioned`

File: controllers/metal3.io/host_state_machine.go
Comment (andfasano) [inline controllers/metal3.io/host_state_machine.go:466]: Looks like this one should be already covered by `imageProvisioningCancelled()`?

File: pkg/provisioner/ironic/ironic.go
Comment (andfasano) [inline pkg/provisioner/ironic/ironic.go:224]: Question: could be this one considered a potential breaking change worth to be documented?

File: pkg/provisioner/ironic/ironic.go
Comment (andfasano) [inline pkg/provisioner/ironic/ironic.go:1347]: Maybe a named const here could be clearer

File: controllers/metal3.io/baremetalhost_controller.go
Comment (zaneb) [inline controllers/metal3.io/baremetalhost_controller.go:832]: nit: I'd suggest putting this right after the Image to keep related data together.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (zaneb) [inline controllers/metal3.io/baremetalhost_controller.go:860]: That last comparison is I think valid because the struct only contains a string, but if it contained anything more complex (and the whole point of defining a struct is to give us an option of adding more complex stuff in future AIUI) then it wouldn't. If we use reflect.DeepEqual() here now we can save someone a headache in future. (Yes, the image comparison above is also dodgy.)

File: controllers/metal3.io/baremetalhost_controller_test.go
Comment (dtantsur) [reply controllers/metal3.io/baremetalhost_controller_test.go:1040]: I'm not sure what you mean by "table driven mode", could you elaborate?

File: pkg/provisioner/ironic/ironic.go
Comment (dtantsur) [reply pkg/provisioner/ironic/ironic.go:224]: Do we have any sort of release notes for that?

File: pkg/provisioner/ironic/ironic.go
Comment (dtantsur) [reply pkg/provisioner/ironic/ironic.go:920]: I'd avoid empty strings, they're not equivalent to `nil`.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [reply controllers/metal3.io/baremetalhost_controller.go:827]: I tried that, and I think you were the one to suggest we don't :) In any case, it's going to be very invasive.

File: apis/metal3.io/v1alpha1/baremetalhost_types.go
Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_types.go:757]: Not too much? We'd still have to deal with Image.URL possibly be "", CustomDeploy possibly be ""/`nil` and so on. A potential user confusion because of conflating too barely related terms may not be worth it.

File: controllers/metal3.io/host_state_machine.go
Comment (dtantsur) [reply controllers/metal3.io/host_state_machine.go:466]: It's not. The two functions are called with `&&` between them, so if `imageProvisioningCancelled` returns `true`, this will return `false` and the result will be `false`.

File: controllers/metal3.io/baremetalhost_controller_test.go
Comment (andfasano) [reply controllers/metal3.io/baremetalhost_controller_test.go:1040]: Yes, sure, here's an example: https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/updatehardwarestate_test.go.

Essentially in the test we're going to define a number of different relevant cases (each one with its own inputs and expectation), and such approach helps in reducing the duplications, reduce the risks of side effects, and definitely supports troubleshooting and maintenance.

(there's a great blog about it here https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)

File: pkg/provisioner/ironic/ironic.go
Comment (andfasano) [reply pkg/provisioner/ironic/ironic.go:225]: AFAIK I don't think we have rl, and probably this could be a good point. Currently, there's just the docs in the `\docs` folder, a possibility could be to add a `Requirements` section there

File: controllers/metal3.io/host_state_machine.go
Comment (andfasano) [reply controllers/metal3.io/host_state_machine.go:466]: Yes, that's right, but it looks a little bit tricky, and since this method could trigger a disruptive  action like deprovisioning I think it's worth trying to simplify it as much as possible.
I guess that the goal of the `provisioningCancelled()` method is essentially to detect a change - during the `Provisioning` or `Provisioned` state - such that invalidates the current state, thus requiring a deprovisioning.

As per the proposal, the  `Image` field should be considered optional if `CustomDeploy` is set (but not if it isn't of course), and this may suggest to:
1)  Check first `CustomDeploy` and then `Image` 
    * Do not consider `Image` nullity as a trigger condition in such case if `CustomDeploy` is set
2) If `CustomDeploy` is nil, check `Image` field as before to ensure the previous behavior

Ie, something like that: 

```
func (hsm *hostStateMachine) provisioningCancelled() bool {

	if hsm.Host.Spec.CustomDeploy != nil {
		if hsm.Host.Spec.CustomDeploy.Method == "" {
			return true
		}
		if hsm.Host.Status.Provisioning.CustomDeploy != nil && hsm.Host.Spec.CustomDeploy.Method != hsm.Host.Status.Provisioning.CustomDeploy.Method {
			return true
		}
		// At this point spec.CustomDeploy value didn't change (and it's not empty). Only a discrepancy in the Image
		// could require a deprovisioning, but if it isn't set or empty then is fine
		if hsm.Host.Spec.Image != nil && hsm.Host.Status.Provisioning.Image.URL != "" && hsm.Host.Spec.Image.URL != hsm.Host.Status.Provisioning.Image.URL {
			return true
		}

		return false
	}

	// Fallback to the previous case
	return hsm.imageProvisioningCancelled()
}
```

Comment (zaneb) [reply controllers/metal3.io/host_state_machine.go:466]: I also read this wrong, and I have created truly terrible bugs in the past (see #915) by getting this function wrong, so I agree that we should seek to simplify the logic.

File: pkg/provisioner/ironic/ironic.go
Comment (zaneb) [reply pkg/provisioner/ironic/ironic.go:921]: To be clearer, there are theoretically 3 ways to clear a field in Ironic, only 2 of which are possible using SetInstanceInfoOpts:

* Update the field and set the value to `""` - passing `""` does this as you'd expect
* Update the field and set the value to `null` - there's no way to do this
* Delete the field - passing `nil` will do this

Sounds like we do want the latter though, when the return from GetChecksum() is false.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (zaneb) [reply controllers/metal3.io/baremetalhost_controller.go:831]: Oh, yeah I guess it's probably even more messy on the provisioner side to flip it.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (zaneb) [inline controllers/metal3.io/baremetalhost_controller.go:555]: I think the correct logic here is:

```
if host.Status.Provisioning.CustomDeploy != nil && host.Status.Provisioning.CustomDeploy.Method != "" {
    return true
}

switch host.Status.Provisioning.State {
case metal3v1alpha1.StateProvisioning, metal3v1alpha1.StateExternallyProvisioned:
    return host.Spec.CustomDeploy != nil && host.Spec.CustomDeploy.Method != ""
default:
    return false
}
```

Comment (dtantsur) [reply controllers/metal3.io/baremetalhost_controller.go:555]: Thank you!

File: controllers/metal3.io/host_state_machine.go
Comment (dtantsur) [reply controllers/metal3.io/host_state_machine.go:466]: Will do, but the proposed snippet doesn't cover one case (removing custom deploy while keeping the image). I'll try to figure it out.

File: pkg/provisioner/ironic/ironic.go
Comment (dtantsur) [reply pkg/provisioner/ironic/ironic.go:225]: I'll put something in configuration.md

File: pkg/provisioner/ironic/ironic.go
Comment (dtantsur) [reply pkg/provisioner/ironic/ironic.go:921]: Aha, got it.

**Comment by dtantsur** [comment]
/test-integration 

All comments addressed, please check again.

**Comment by andfasano** [comment]
Thanks @dtantsur, current improvements seem fine to me.

/lgtm

**Comment by s3rj1k** [comment]
@dtantsur What is the minimal Ironic release after this PR gets merged, also can this be stated somewhere in docs?
Thanks for your hard work.

**Comment by dtantsur** [comment]
> What is the minimal Ironic release after this PR gets merged, also can this be stated somewhere in docs?

We're stating the minimum API version in configuration.md, which can be matched to the actual service version in https://docs.openstack.org/ironic/latest/contributor/webapi-version-history.html. This requires 1.62 (Wallaby).

File: controllers/metal3.io/host_state_machine.go
Comment (zaneb) [inline controllers/metal3.io/host_state_machine.go:432]: Not sure why an empty custom deploy struct causes provisioning to be unconditionally cancelled? I think that this should go away and the line above should be changed to:

```
if hsm.Host.Spec.CustomDeploy != nil && hsm.Host.Spec.CustomDeploy.Method != "" {
```

In my view the only relevant thing is the Method, and the checks for nil are only to prevent crashes not something that should provide a subtle user interface to behaviour changes.

File: controllers/metal3.io/host_state_machine.go
Comment (zaneb) [inline controllers/metal3.io/host_state_machine.go:440]: I think this should be:

```
if hsm.Host.Status.Provisioning.Image.URL != "" &&
    (hsm.Host.Spec.Image == nil ||
        hsm.Host.Spec.Image.URL != hsm.Host.Status.Provisioning.Image.URL) {
```

File: controllers/metal3.io/host_state_machine.go
Comment (zaneb) [inline controllers/metal3.io/host_state_machine.go:449]: nit: 'previous' here refers to history not to control flow, so it will make less sense outside the context of the code review.

File: controllers/metal3.io/host_state_machine.go
Comment (zaneb) [inline controllers/metal3.io/host_state_machine.go:432]: The `Status.Provisioning.CustomDeploy.Method` is expected to be empty during provisioning. I think _in practice_ that `Status.Provisioning.CustomDeploy` will be nil also, but it seems like it would be more robust to check both:

```
if hsm.Host.Status.Provisioning.CustomDeploy != nil &&
    hsm.Host.Status.Provisioning.CustomDeploy.Method != "" &&
        hsm.Host.Spec.CustomDeploy.Method != hsm.Host.Status.Provisioning.CustomDeploy.Method {
```

Comment (dtantsur) [reply controllers/metal3.io/host_state_machine.go:432]: May I follow-up with this? I'd prefer to avoid another round of looking for lgtm and fighting the CI because of something so minor.

Comment (dtantsur) [reply controllers/metal3.io/host_state_machine.go:432]: https://github.com/metal3-io/baremetal-operator/pull/932

**Pull request merged**
